src/multiversioning.zig

```diff
@@ -0,0 +1,836 @@
+        return writer.print("{}.{}.{}", .{
+            release_triple.major,
+            release_triple.minor,
+            release_triple.patch,
+        });
+    }
+
+    pub fn max(a: Release, b: Release) Release {
+        if (a.value > b.value) {
+            return a;
+        } else {
+            return b;
+        }
+    }
+};
+
+pub const ReleaseTriple = extern struct {
+    patch: u8,
+    minor: u8,
+    major: u16,
+
+    comptime {
+        assert(@sizeOf(ReleaseTriple) == 4);
+        assert(stdx.no_padding(ReleaseTriple));
+    }
+
+    pub fn parse(string: []const u8) error{InvalidRelease}!ReleaseTriple {
+        var parts = std.mem.splitScalar(u8, string, '.');
+        const major = parts.first();
+        const minor = parts.next() orelse return error.InvalidRelease;
+        const patch = parts.next() orelse return error.InvalidRelease;
+        if (parts.next() != null) return error.InvalidRelease;
+        return .{
+            .major = std.fmt.parseUnsigned(u16, major, 10) catch return error.InvalidRelease,
+            .minor = std.fmt.parseUnsigned(u8, minor, 10) catch return error.InvalidRelease,
+            .patch = std.fmt.parseUnsigned(u8, patch, 10) catch return error.InvalidRelease,
+        };
+    }
+};
+
+test "ReleaseTriple.parse" {
+    const tests = [_]struct {
+        string: []const u8,
+        result: error{InvalidRelease}!ReleaseTriple,
+    }{
+        // Valid:
+        .{ .string = "0.0.1", .result = .{ .major = 0, .minor = 0, .patch = 1 } },
+        .{ .string = "0.1.0", .result = .{ .major = 0, .minor = 1, .patch = 0 } },
+        .{ .string = "1.0.0", .result = .{ .major = 1, .minor = 0, .patch = 0 } },
+
+        // Invalid characters:
+        .{ .string = "v0.0.1", .result = error.InvalidRelease },
+        .{ .string = "0.0.1v", .result = error.InvalidRelease },
+        // Invalid separators:
+        .{ .string = "0.0.0.1", .result = error.InvalidRelease },
+        .{ .string = "0..0.1", .result = error.InvalidRelease },
+        // Overflow (and near-overflow):
+        .{ .string = "0.0.255", .result = .{ .major = 0, .minor = 0, .patch = 255 } },
+        .{ .string = "0.0.256", .result = error.InvalidRelease },
+        .{ .string = "0.255.0", .result = .{ .major = 0, .minor = 255, .patch = 0 } },
+        .{ .string = "0.256.0", .result = error.InvalidRelease },
+        .{ .string = "65535.0.0", .result = .{ .major = 65535, .minor = 0, .patch = 0 } },
+        .{ .string = "65536.0.0", .result = error.InvalidRelease },
+    };
+    for (tests) |t| {
+        try std.testing.expectEqualDeep(ReleaseTriple.parse(t.string), t.result);
+    }
+}
+
+pub const MultiversionMetadata = extern struct {
+    // When slicing into the binary:
+    // checksum(section[past_offset..past_offset+past_size]) == past_checksum.
+    // This is then validated when the binary is written to a memfd or similar.
+    // TODO: Might be nicer as an AoS? It's control plane state.
```

Comment:
> Checking my understanding: We could change this MultiversionMetadata schema later, and to transition the multi-version build step would need to decode the old version and encode the new version -- at which point we would not need to old version again.

Yes, that's correct. Much in the same way the epoch is currently a special case.

The other consideration is that this struct is read by past versions of TigerBeetle, we have 2 escape hatches there:
* If we really need to break compatibility, the upgrade process becomes "upgrade the binary, restart each replica"
* It's always the latest version that does the reading and executing, so otherwise the information is used for advertising. So, if we had to add compression, we wouldn't need to change anything in a breaking way here. 

  As long as the top level checksums (`checksum_header`, `checksum_binary_without_header`), and `past.count`, `past.versions` and `past.visits` are readable, the old version will correctly advertise the new version, which can then do anything with the rest of the fields (or additional ones) as it needs.

  This makes it quite flexible! I'm thinking of dropping the explicit `reserved` and check for `elf_section_header.sh_size != @sizeOf(MultiversionHeader)` because we can add fields freely if we need with this approach.